### DISCUSSION TITLE
Missing a directory

### DISCUSSION THREAD

**Discussion by jamescampagnola** [community/community#55421] [New to GitHub]
When I open my codespace I see a folder in the explorer tab on the left side.  Then when I go to Github.com and view my profile, one of the folders is missing.  Anyone have any ideas about this.  Thanks very much.  Jim

**Comment by jge162** [comment]
Why don’t you just do `git pull origin main` to update remote repository. So codespaces and GitHub repository match.

**Comment by airtower-luna** [comment]
Did you commit and push files in that folder? It's not going to end up in your repository without that.

**Comment by jge162** [comment]
> When I open my codespace I see a folder in the explorer tab on the left side. Then when I go to Github.com and view my profile, one of the folders is missing. Anyone have any ideas about this. Thanks very much. Jim

Do this

`git pull origin main`
`git add .`
`git commit -m “your message here”`
`git push origin main`

**Comment by jamescampagnola** [comment]
Thanks very much folks for the quick repsonse and your help. Apologies if I'm in the wrong place but I posted to the the 'New To Github' group because I'm a serious beginner.  So basically, when I go to github.com I'm taken to what looks like my 'homepage.' On the left column it says 'Top Repositories'. When I click on the 'repository' I'm working in, which is a CS50 codespace, I'm taken to a page that shows what looks like a directory of folders I've created in the codespace 'explorer'. Is this a 'Remote Repositoty'? But this directory is missing a folder that I'm currently working in and which I see when I launch my codespace.  In addition, I'm periodically getting a message in the codespace terminal window- "Remote repository contains conflicting changes", with a "Force Push"  button. The "Force Push" button doesn't seem to reslove anything.  Since 'git pull origin main' looks like a terminal command, probably not a great idea to start tooling around in the terminal until I understand what 'pull', 'commit' and 'push' commands are. If anyone understands what I'm describing or can recommend a tutorial or where I can get some basic info, I'd really appreciate it.  I haven't figured out what the training slideshow in my list of repositories does or how it works.  It might need its own training tutorial.  Thanks again...

**Reply by jge162** [reply]
Just keep in mind if the codespace does not match all the files in your repository that’s not a big deal. Just delete it and create a new codespace. You use the git commands to update the local repository when you want the pushes from codespaces (remote) changes to be updated in your main repo. 

I use codespaces for web development to test web build. Then when I am happy I git add, then commit and finally push from remote to local repo so website will show updated changes I made in codespaces. 

in short, For example if you totally mess up the remote codespaces not a big deal. Just delete and start over. It will build based off the main repo.

**Reply by jge162** [reply]
You also should get familiar with git commands as it’s the best way to push changes from remote to local repo. Practice makes perfect. Don’t be afraid. Just don’t push to main anything you don’t need in main. But play with it.

**Reply by jge162** [reply]
Just do git pull origin main. That makes sure that all the stuff in main repo is pulled into the remote.

**Reply by jge162** [reply]
> Thanks very much folks for the quick repsonse and your help. Apologies if I'm in the wrong place but I posted to the the 'New To Github' group because I'm a serious beginner. So basically, when I go to github.com I'm taken to what looks like my 'homepage.' On the left column it says 'Top Repositories'. When I click on the 'repository' I'm working in, which is a CS50 codespace, I'm taken to a page that shows what looks like a directory of folders I've created in the codespace 'explorer'. Is this a 'Remote Repositoty'? But this directory is missing a folder that I'm currently working in and which I see when I launch my codespace. In addition, I'm periodically getting a message in the codespace terminal window- "Remote repository contains conflicting changes", with a "Force Push" button. The "Force Push" button doesn't seem to reslove anything. Since 'git pull origin main' looks like a terminal command, probably not a great idea to start tooling around in the terminal until I understand what 'pull', 'commit' and 'push' commands are. If anyone understands what I'm describing or can recommend a tutorial or where I can get some basic info, I'd really appreciate it. I haven't figured out what the training slideshow in my list of repositories does or how it works. It might need its own training tutorial. Thanks again...

in regards to force push, if you did git pull origin main first this would not show up.

**Reply by jamescampagnola** [reply]
Thanks jge162. I guess the first problem is that I don't know the difference between 'local' and 'remote.'  I don't have Git installed on my computer and when I launch the codespace, it's in a browser.  So when I launch my codespace in the browser, am I looking at a 'local' or a 'remote' codespace?  And if I delete the codespace in the browser, is that a 'local' codespace? And where can I find the 'main' repo?  Is that remote? Where does it live?

**Reply by jge162** [reply]
Local will be considered the repo on GitHub.com

codespace is a remote space where you can edit the files you pulled from the repo  at GitHub.com when the codespace was created. 

just open a terminal window on codespace and you can use git there. On the webpage it loads too. 

Key board options are (Control and  ~ ) will load terminal in codespace. From there you can use git commands.

**Reply by jge162** [reply]
@jamescampagnola here are two screen shots, first showing how you can delete a codespace, remote repo. then the second the process for updating the remote repo first with a pull request. then second staging the changed for push from remote to local repo.

<img width="394" alt="Screenshot 2023-05-14 at 3 00 24 PM" src="https://github.com/community/community/assets/31228460/11021d88-1137-46c5-b8c5-31728499b4ea">
<img width="577" alt="Screenshot 2023-05-14 at 3 01 40 PM" src="https://github.com/community/community/assets/31228460/7177abe2-7193-47cf-b0b4-28e1cda885af">

**Reply by jamescampagnola** [reply]
Thanks.  The first screenshot, where is that? What's the url? I guess it's a page in an account? Not sure how to get there..

**Reply by jge162** [reply]
it is the main page of the the repository , under the section code.

<img width="1021" alt="Screenshot 2023-05-14 at 3 46 53 PM" src="https://github.com/community/community/assets/31228460/8835fc48-ffc5-4d55-a7a3-5d75f3ca8acc">

**Comment by airtower-luna** [comment]
For a beginner, the most important part is probably to understand what Git is and how it works. Check the "Git Book", the first three chapters cover the basics that should let you work: https://git-scm.com/book/en/v2

For the confusion about different files in your GitHub repository and codespace, it's important to understand that you're not looking at the same thing. The codespace has its own copy of the repository. If you want to get a change you made in the codespace to show up on GitHub, you need to first *commit* it (tell Git to record the change), then *push* the new commit to the repository on GitHub. Likewise if you want changes on GitHub to appear in your codespace, you'll need to *pull* them in.

All those concepts are explained in the book (which is a website, though you can download an ebook version if you want). And they work the same whether you work in a codespace or have Git right on your computer. :slightly_smiling_face:

**Reply by jamescampagnola** [reply]
Thanks airtower-luna.  I'll get on that book tonight. So if I'm working in the codespace, how do I tell Git to commit the changes?  I thought that was happening automatically...

**Reply by airtower-luna** [reply]
With the `git commit` command, similar to the screenshot @jge162 posted above. It's explained in the Git Book here: [Git Basics - Recording Changes to the Repository](https://git-scm.com/book/en/v2/Git-Basics-Recording-Changes-to-the-Repository)

It's possible the Codespaces UI has some menu to do the same, but I'm not familiar with it.
